src/Language/Futhark/Interpreter.hs

```diff
@@ -1921,11 +1984,90 @@ initialCtx =
           else pure $ toArray shape $ map (toArray rowshape) $ chunk (asInt m) xs'
     def "manifest" = Just $ fun1 pure
     def "vjp2" = Just $
-      fun3 $
-        \_ _ _ -> bad noLoc mempty "Interpreter does not support autodiff."
+      fun3 $ \f v s -> do
+        -- Get the depth
+        depth <- length <$> stacktrace
+
+        -- Get the seed values
+        let getSeeds v = case v of
+              ValuePrim _ -> [valueToAD v]
+              ValueSeed _ -> [valueToAD v]
+              ValueRecord m -> foldl1 (++) (M.map getSeeds m)
+              _ -> error "Not implemented (d19h45iu782)" -- TODO: Arrays?
+        let s' = getSeeds s
+
+        -- Impregnate the values
+        let addSeeds i v = case v of
+              ValuePrim v' -> (i + 1, ValueSeed $ AD.VjpSeed depth $ AD.TapeId i $ AD.Primal $ fromJust $ getV v')
+              ValueSeed v' -> (i + 1, ValueSeed $ AD.VjpSeed depth $ AD.TapeId i $ AD.Seed v')
+              ValueRecord m -> second ValueRecord $ M.mapAccum addSeeds i m
+              _ -> error "Not implemented (gs3g3ss)" -- TODO: Arrays?
```

Comment:
In order to improve the way this is written (and to handle arrays), I suggest defining a utility function for performing a monadic traversal across the "leaves" of an interpreter value. Maybe something with this type:

```Haskell
traverseValue :: Monad f => (PrimValue -> f (Value m)) -> (AD.ADSeed -> f (Value m)) -> Value m -> f (Value m)
```